src/stream/src/executor/sort_buffer.rs

```diff
@@ -11,385 +11,96 @@
-/// with a delay.
 pub struct SortBuffer<S: StateStore> {
-    schema: Schema,
-
-    pk_indices: PkIndices,
-
-    sort_column_index: usize,
-
-    state_table: StateTable<S>,
-
-    chunk_size: usize,
+    _phantom: PhantomData<S>,
+}
 
-    last_output: Option<ScalarImpl>,
+impl<S: StateStore> SortBuffer<S> {
+    pub fn new() -> Self {
+        Self {
+            _phantom: PhantomData,
+        }
+    }
 
-    watermark: Option<ScalarImpl>,
+    /// Insert a new row into the buffer.
+    pub fn insert(&mut self, new_row: impl Row, buffer_table: &mut StateTable<S>) {
+        buffer_table.insert(new_row);
+    }
 
-    buffer: BTreeMap<SortBufferKey, SortBufferValue>,
-}
+    /// Delete a row from the buffer.
+    pub fn delete(&mut self, old_row: impl Row, buffer_table: &mut StateTable<S>) {
+        buffer_table.delete(old_row);
+    }
 
-impl<S: StateStore> SortBuffer<S> {
-    /// Update the watermark value.
-    pub fn set_watermark(&mut self, watermark: ScalarImpl) {
-        assert!(
-            {
-                if let Some(ref last_watermark) = self.watermark {
-                    &watermark >= last_watermark
-                } else {
-                    true
-                }
-            },
-            "watermark should always increase"
-        );
-        self.watermark = Some(watermark);
+    /// Update a row in the buffer.
+    pub fn update(
+        &mut self,
+        old_row: impl Row,
+        new_row: impl Row,
+        buffer_table: &mut StateTable<S>,
+    ) {
+        buffer_table.update(old_row, new_row);
     }
 
-    pub async fn recover(
-        schema: Schema,
-        pk_indices: PkIndices,
-        sort_column_index: usize,
-        state_table: StateTable<S>,
-    ) -> StreamExecutorResult<Self> {
-        let vnodes = state_table.vnode_bitmap().to_owned();
-        let mut buffer = BTreeMap::new();
+    /// Apply a change to the buffer, insert/delete/update.
+    pub fn apply_change(&mut self, change: &Record<OwnedRow>, buffer_table: &mut StateTable<S>) {
+        match change {
+            Record::Insert { new_row } => self.insert(new_row, buffer_table),
+            Record::Delete { old_row } => self.delete(old_row, buffer_table),
+            Record::Update { old_row, new_row } => self.update(old_row, new_row, buffer_table),
+        }
+    }
 
+    /// Consume rows under `watermark` from the buffer.
+    /// NOTICE: state in `buffer_table` must be correctly cleaned after calling this method.
+    #[try_stream(ok = OwnedRow, error = StreamExecutorError)]
+    pub async fn consume<'a>(
+        &'a mut self,
+        watermark: ScalarImpl,
+        buffer_table: &'a mut StateTable<S>,
+    ) {
         let pk_range = (
             Bound::<row::Empty>::Unbounded,
-            Bound::<row::Empty>::Unbounded,
+            // TODO(): include or exclude?
+            Bound::Included([Some(watermark.as_scalar_ref_impl())]),
```

Comment:
> Is it because the `delete_ranges` in state table are left-inclusive-right-exclusive?

No. It is because the definition of watermark w is [w, +inf)

Reply:
Okay